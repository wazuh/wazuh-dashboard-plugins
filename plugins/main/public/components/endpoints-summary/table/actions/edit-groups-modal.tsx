import React, { useState, useEffect } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiBadge,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';
import { useGetGroups } from '../../hooks';
import {
  addAgentToGroupService,
  removeAgentFromGroupsService,
} from '../../services';
import { Agent } from '../../types';
import { getToasts } from '../../../../kibana-services';

interface EditAgentGroupsModalProps {
  agent: Agent;
  onClose: () => void;
  reloadAgents: () => void;
}

export const EditAgentGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(({ agent, onClose, reloadAgents }: EditAgentGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState(
    agent.group?.map(group => ({ label: group })) || [],
  );
  const [isSaving, setIsSaving] = useState(false);

  const {
    groups,
    isLoading: isGroupsLoading,
    error: errorGroups,
  } = useGetGroups();

  if (errorGroups) {
    const options = {
      context: `EditAgentGroupsModal.useGetGroups`,
      level: UI_LOGGER_LEVELS.ERROR,
      severity: UI_ERROR_SEVERITIES.BUSINESS,
      store: true,
      error: {
        error: errorGroups,
        message: errorGroups.message || errorGroups,
        title: `Could not get groups`,
      },
    };
    getErrorOrchestrator().handleError(options);
  }

  const showToast = (
    color: string,
    title: string = '',
    text: string = '',
    time: number = 3000,
  ) => {
    getToasts().add({
      color: color,
      title: title,
      text: text,
      toastLifeTimeMs: time,
    });
  };

  const handleOnSave = async () => {
    setIsSaving(true);

    const flatSelectedGroups = selectedGroups.map(group => group.label);
    const addedGroups =
      flatSelectedGroups.filter(group => !agent.group?.includes(group)) || [];
    const removedGroups =
      agent.group?.filter(group => !flatSelectedGroups.includes(group)) || [];

    if (!removedGroups.length && !addedGroups.length) {
      setIsSaving(false);
      onClose();
      return;
    }

    try {
      addedGroups.length &&
        (await Promise.all(
          addedGroups?.map(groupId =>
            addAgentToGroupService({ agentId: agent.id, groupId }),
          ),
        ));
      removedGroups.length &&
        (await removeAgentFromGroupsService({
          agentId: agent.id,
          groupIds: removedGroups,
        }));
      showToast('success', 'Edit agent groups', 'Groups saved successfully');
      reloadAgents();
    } catch (error) {
      const options = {
        context: `EditAgentGroupsModal.handleOnSave`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error,
          message: error.message || error,
          title: `Could not save agent groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    } finally {
      setIsSaving(false);
      onClose();
    }
  };

  const form = (
    <EuiForm component='form'>
      <EuiFormRow label='Agent'>
        <EuiBadge color='hollow'>{agent.name}</EuiBadge>
      </EuiFormRow>
      <EuiFormRow
        label='Groups'
        isRequired
        isInvalid={!selectedGroups?.length}
        error={['You must add at least one group']}
      >
        <EuiComboBox
          placeholder='Select groups'
          options={groups?.map(group => ({ label: group })) || []}
          selectedOptions={selectedGroups}
          onChange={selectedGroups => setSelectedGroups(selectedGroups)}
          isLoading={isGroupsLoading}
          clearOnBlur
        />
      </EuiFormRow>
    </EuiForm>
  );

  return (
    <EuiModal
      onClose={onClose}
      onClick={ev => {
        ev.stopPropagation();
      }}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>Edit agent groups</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>{form}</EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
        <EuiButton
          onClick={handleOnSave}
          fill
          isLoading={isSaving}
          disabled={isGroupsLoading || !selectedGroups?.length}
        >
          Save
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
});
