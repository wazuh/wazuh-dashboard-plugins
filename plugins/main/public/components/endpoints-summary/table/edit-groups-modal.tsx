import React, { useState, useEffect } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../react-services/common-services';
import { useGetGroups } from '../hooks/groups';
import {
  addAgentToGroupService,
  removeAgentFromGroupsService,
} from '../services';

interface EditAgentGroupsModalProps {
  agent: { id: string; name: string; group: string[] };
  onClose: () => void;
  reloadAgents: () => void;
}

export const EditAgentGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(({ agent, onClose, reloadAgents }: EditAgentGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState(
    agent.group?.map(group => ({ label: group })) || [],
  );
  const [isSaving, setIsSaving] = useState(false);

  const {
    groups,
    isLoading: isGroupsLoading,
    error: errorGroups,
  } = useGetGroups();

  if (errorGroups) {
    const options = {
      context: `EditAgentGroupsModal.useGetGroups`,
      level: UI_LOGGER_LEVELS.ERROR,
      severity: UI_ERROR_SEVERITIES.BUSINESS,
      store: true,
      error: {
        error,
        message: error.message || error,
        title: `Could not get groups`,
      },
    };
    getErrorOrchestrator().handleError(options);
  }

  const handleOnSave = async () => {
    setIsSaving(true);

    const flatSelectedGroups = selectedGroups.map(group => group.label);
    const addedGroups =
      flatSelectedGroups.filter(group => !agent.group?.includes(group)) || [];
    const removedGroups =
      agent.group?.filter(group => !flatSelectedGroups.includes(group)) || [];

    if (!removedGroups.length && !addedGroups.length) {
      setIsSaving(false);
      onClose();
      return;
    }

    try {
      addedGroups.length &&
        (await Promise.all(
          addedGroups?.map(group => addAgentToGroupService(agent.id, group)),
        ));
      removedGroups.length &&
        (await removeAgentFromGroupsService(agent.id, removedGroups));
      reloadAgents();
    } catch (error) {
      const options = {
        context: `EditAgentGroupsModal.handleOnSave`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error,
          message: error.message || error,
          title: `Could not save agent groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    } finally {
      setIsSaving(false);
      onClose();
    }
  };

  const form = (
    <EuiForm component='form'>
      <EuiFormRow label='Agent'>
        <EuiText>
          <b>{agent.name}</b>
        </EuiText>
      </EuiFormRow>
      <EuiFormRow
        label='Groups'
        isRequired
        isInvalid={!selectedGroups?.length}
        error={['You must add at least one group']}
      >
        <EuiComboBox
          placeholder='Select groups'
          options={groups?.map(group => ({ label: group })) || []}
          selectedOptions={selectedGroups}
          onChange={selectedGroups => setSelectedGroups(selectedGroups)}
          isLoading={isGroupsLoading}
        />
      </EuiFormRow>
    </EuiForm>
  );

  return (
    <EuiModal
      onClose={onClose}
      initialFocus='[name=popswitch]'
      onClick={ev => {
        ev.stopPropagation();
      }}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>Edit agent groups</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>{form}</EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
        <EuiButton
          onClick={handleOnSave}
          fill
          isLoading={isSaving}
          disabled={isGroupsLoading || !selectedGroups?.length}
        >
          Save
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
});
