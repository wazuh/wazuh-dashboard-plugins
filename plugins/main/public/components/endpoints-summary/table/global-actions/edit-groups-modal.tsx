import React, { useState } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
} from '@elastic/eui';
import { compose } from 'redux';
import { withErrorBoundary, withReduxProvider } from '../../../common/hocs';
import { UI_LOGGER_LEVELS } from '../../../../../common/constants';
import { UI_ERROR_SEVERITIES } from '../../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../../react-services/common-services';
import { useGetGroups } from '../../hooks';
import {
  addAgentToGroupService,
  removeAgentFromGroupService,
} from '../../services';
import { Agent } from '../../types';
import {
  AgentsGlobalActionsResultModal,
  RESULT_TYPE,
  Result,
} from './result-modal';

interface EditAgentsGroupsModalProps {
  agents: Agent[];
  onClose: () => void;
  reloadAgents: () => void;
  addOrRemove: 'add' | 'remove';
}

type Option = {
  label: string;
};

export const EditAgentsGroupsModal = compose(
  withErrorBoundary,
  withReduxProvider,
)(
  ({
    agents,
    onClose,
    reloadAgents,
    addOrRemove,
  }: EditAgentsGroupsModalProps) => {
    const [selectedGroups, setSelectedGroups] = useState<Option[]>([]);
    const [isSaving, setIsSaving] = useState(false);
    const [isResultModalVisible, setIsResultModalVisible] = useState(false);
    const [results, setResults] = useState<Result[]>([]);

    const {
      groups,
      isLoading: isGroupsLoading,
      error: errorGroups,
    } = useGetGroups();

    if (errorGroups) {
      const options = {
        context: `EditAgentsGroupsModal.useGetGroups`,
        level: UI_LOGGER_LEVELS.ERROR,
        severity: UI_ERROR_SEVERITIES.BUSINESS,
        store: true,
        error: {
          error: errorGroups,
          message: errorGroups.message || errorGroups,
          title: `Could not get groups`,
        },
      };
      getErrorOrchestrator().handleError(options);
    }

    const getStringArrayGroups = (options: Option[]) => {
      return options.map(group => group.label);
    };

    const handleOnChange = (selectedGroups: Option[]) => {
      setSelectedGroups(selectedGroups);
    };

    const handleOnSave = async () => {
      setIsSaving(true);

      try {
        const promises = agents.reduce((acc, agent) => {
          const groups = getStringArrayGroups(selectedGroups);

          if (addOrRemove === 'add') {
            const groupsToAdd = groups.filter(
              group => !agent.group?.includes(group),
            );

            const groupsToAddPromises = groupsToAdd?.map(group =>
              addAgentToGroupService(agent.id, group)
                .then(() =>
                  setResults(results => {
                    const newResult = {
                      type: RESULT_TYPE.SUCCESS,
                      agent,
                      group,
                    };
                    return [...results, newResult];
                  }),
                )
                .catch(error => {
                  setResults(results => {
                    const newResult = {
                      type: RESULT_TYPE.ERROR,
                      agent,
                      group,
                      message: error.message,
                    };
                    return [...results, newResult];
                  });
                }),
            );

            return [...acc, ...groupsToAddPromises];
          }

          const groupsToRemove = groups.filter(group =>
            agent.group?.includes(group),
          );

          const groupsToRemovePromises = groupsToRemove?.map(group =>
            removeAgentFromGroupService(agent.id, group)
              .then(() =>
                setResults(results => {
                  const newResult = {
                    type: RESULT_TYPE.SUCCESS,
                    agent,
                    group,
                  };
                  return [...results, newResult];
                }),
              )
              .catch(error => {
                setResults(results => {
                  const newResult = {
                    type: RESULT_TYPE.ERROR,
                    agent,
                    group,
                    message: error.message,
                  };
                  return [...results, newResult];
                });
              }),
          );

          return [...acc, ...groupsToRemovePromises];
        }, [] as any);

        await Promise.allSettled(promises);

        setIsSaving(false);
        setIsResultModalVisible(true);

        reloadAgents();
      } catch (error) {
        setIsSaving(false);
        const options = {
          context: `EditAgentsGroupsModal.handleOnSave`,
          level: UI_LOGGER_LEVELS.ERROR,
          severity: UI_ERROR_SEVERITIES.BUSINESS,
          store: true,
          error: {
            error,
            message: error.message || error,
            title: `Could not save agents groups`,
          },
        };
        getErrorOrchestrator().handleError(options);
      }
    };

    const groupsText =
      addOrRemove === 'add'
        ? 'Select groups to add'
        : 'Select groups to remove';

    const form = (
      <EuiForm component='form'>
        <EuiFormRow label='Seleted agents'>
          <EuiText>{agents.length}</EuiText>
        </EuiFormRow>
        <EuiFormRow label={groupsText}>
          <EuiComboBox
            placeholder={groupsText}
            options={groups?.map(group => ({ label: group })) || []}
            selectedOptions={selectedGroups}
            onChange={handleOnChange}
            isLoading={isGroupsLoading}
            clearOnBlur
          />
        </EuiFormRow>
      </EuiForm>
    );

    return (
      <EuiModal onClose={onClose}>
        {isResultModalVisible ? (
          <AgentsGlobalActionsResultModal
            title={
              addOrRemove === 'add'
                ? 'Add groups to agents'
                : 'Remove groups from agents'
            }
            isLoading={isSaving}
            results={results}
            onClose={onClose}
            loadingMessage='Adding groups to agents...'
          />
        ) : (
          <>
            <EuiModalHeader>
              <EuiModalHeaderTitle>
                {addOrRemove === 'add'
                  ? 'Add groups to agents'
                  : 'Remove groups from agents'}
              </EuiModalHeaderTitle>
            </EuiModalHeader>
            <EuiModalBody>{form}</EuiModalBody>
            <EuiModalFooter>
              <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
              <EuiButton
                onClick={handleOnSave}
                fill
                isLoading={isSaving}
                disabled={isGroupsLoading || !selectedGroups?.length}
              >
                Save
              </EuiButton>
            </EuiModalFooter>
          </>
        )}
      </EuiModal>
    );
  },
);
