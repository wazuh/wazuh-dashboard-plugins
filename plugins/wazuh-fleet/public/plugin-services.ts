import { CoreStart } from 'opensearch-dashboards/public';
import { createGetterSetter } from '../../../src/plugins/opensearch_dashboards_utils/common';
import { WazuhCorePluginStart } from '../../wazuh-core/public';
import { AppPluginStartDependencies } from './types';
import { AppSetup } from './application/types';

export const [getPlugins, setPlugins] =
  createGetterSetter<AppPluginStartDependencies>('Plugins');
export const [getCore, setCore] = createGetterSetter<CoreStart>('Core');
export const [getWazuhCore, setWazuhCore] =
  createGetterSetter<WazuhCorePluginStart>('WazuhCore');
export const [getEnrollAgentManagement, setEnrollAgentManagement] =
  createGetterSetter<AppSetup['enrollmentAgentManagement']>(
    'fleetManagementEnrollmentAgentManagement',
  );
