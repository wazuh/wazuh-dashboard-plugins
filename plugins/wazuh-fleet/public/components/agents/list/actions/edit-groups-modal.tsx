import React, { useState } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiFlexGroup,
  EuiFlexItem,
  EuiDescriptionList,
  EuiDescriptionListTitle,
  EuiDescriptionListDescription,
} from '@elastic/eui';
import { getAgentManagement, getToasts } from '../../../../plugin-services';
import { IAgentResponse } from '../../../../../common/types';

interface EditAgentGroupsModalProps {
  agent: IAgentResponse;
  onClose: () => void;
  reloadAgents: () => void;
}

export const EditAgentGroupsModal = ({
  agent,
  onClose,
  reloadAgents,
}: EditAgentGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState(
    agent._source.agent.groups?.map(group => ({ label: group })) || [],
  );
  const [isSaving, setIsSaving] = useState(false);

  // Check if selected groups are the same as agent's current groups
  const areGroupsUnchanged = () => {
    const currentGroups = agent._source.agent.groups || [];
    const selectedGroupLabels = selectedGroups.map(group => group.label);

    if (currentGroups.length !== selectedGroupLabels.length) {
      return false;
    }

    // Check if all current groups are in selected groups and vice versa
    return (
      currentGroups.every(group => selectedGroupLabels.includes(group)) &&
      selectedGroupLabels.every(group => currentGroups.includes(group))
    );
  };

  const handleSave = async () => {
    const flatSelectedGroups = selectedGroups.map(group => group.label);
    const addedGroups =
      flatSelectedGroups.filter(
        group => !agent._source.agent.groups?.includes(group),
      ) || [];
    const removedGroups =
      agent._source.agent.groups?.filter(
        group => !flatSelectedGroups.includes(group),
      ) || [];

    // Exit early if no changes
    if (removedGroups.length === 0 && addedGroups.length === 0) {
      setIsSaving(false);
      onClose();

      return;
    }

    try {
      setIsSaving(true);

      // Perform group operations in parallel if needed
      const operations = [];

      if (addedGroups.length > 0) {
        operations.push(
          getAgentManagement().addGroups(agent._source.agent.id, addedGroups),
        );
      }

      if (removedGroups.length > 0) {
        operations.push(
          getAgentManagement().removeGroups({
            agentId: agent._source.agent.id,
            groupIds: removedGroups,
          }),
        );
      }

      await Promise.all(operations);

      getToasts().addInfo({
        title: 'Agent groups edited',
        text: `Agent ${agent._source.agent.name} groups have been updated`,
      });
      setIsSaving(false);
      reloadAgents();
      onClose();
    } catch {
      getToasts().addDanger({
        title: 'Error editing agent groups',
        text: `Failed to update groups for agent ${agent._source.agent.name}`,
      });
      setIsSaving(false);
    }
  };

  const onCreateOption = (searchValue: string) => {
    const newOption = {
      label: searchValue,
    };

    // Select the option.
    setSelectedGroups([...selectedGroups, newOption]);
  };

  const form = (
    <EuiForm component='form'>
      <EuiFlexGroup direction='column' gutterSize='m'>
        <EuiFlexItem>
          <EuiFlexGroup gutterSize='m'>
            <EuiFlexItem>
              <EuiDescriptionList compressed>
                <EuiDescriptionListTitle>Agent ID</EuiDescriptionListTitle>
                <EuiDescriptionListDescription>
                  {agent._source.agent.id}
                </EuiDescriptionListDescription>
              </EuiDescriptionList>
            </EuiFlexItem>
            <EuiFlexItem>
              <EuiDescriptionList compressed>
                <EuiDescriptionListTitle>Agent name</EuiDescriptionListTitle>
                <EuiDescriptionListDescription>
                  {agent._source.agent.name}
                </EuiDescriptionListDescription>
              </EuiDescriptionList>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow
            label='Groups'
            isInvalid={!selectedGroups?.length}
            error={['You must add at least one group']}
          >
            <EuiComboBox
              placeholder='Select groups'
              options={[]}
              selectedOptions={selectedGroups}
              onChange={selectedGroups => setSelectedGroups(selectedGroups)}
              isLoading={false}
              // TODO: Change when the endpoint or index pattern is available to request the groups
              // options={groups?.map(group => ({ label: group })) || []}
              noSuggestions
              onCreateOption={onCreateOption}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiForm>
  );

  return (
    <EuiModal
      onClose={onClose}
      onClick={event => {
        event.stopPropagation();
      }}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>Edit agent groups</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody>{form}</EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
        <EuiButton
          onClick={handleSave}
          fill
          isLoading={isSaving}
          disabled={selectedGroups.length === 0 || areGroupsUnchanged()}
        >
          Save
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
};
