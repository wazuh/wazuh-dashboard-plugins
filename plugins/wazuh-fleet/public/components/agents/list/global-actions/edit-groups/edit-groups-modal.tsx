import React, { useState } from 'react';
import {
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiButtonEmpty,
  EuiButton,
  EuiForm,
  EuiFormRow,
  EuiComboBox,
  EuiText,
  EuiCallOut,
} from '@elastic/eui';
import { IAgentResponse } from '../../../../../../common/types';
import { getAgentManagement } from '../../../../../plugin-services';
import { EditAgentsGroupsModalResult } from './result';
import { EditActionGroups } from './types';

export enum RESULT_TYPE {
  SUCCESS = 'success',
  ERROR = 'error',
}

export interface GroupResult {
  documentId: string;
  result: RESULT_TYPE;
  successAgents?: string[];
  errorMessage?: string;
  totalErrorAgents?: number;
  errorAgents?: [
    {
      error: { message: string };
      id: string;
    },
  ];
}

interface EditAgentsGroupsModalProps {
  selectedAgents: IAgentResponse[];
  allAgentsSelected: boolean;
  params: object;
  onClose: () => void;
  reloadAgents: () => void;
  editAction: EditActionGroups;
}

interface Option {
  label: string;
}

export const EditAgentsGroupsModal = ({
  selectedAgents,
  allAgentsSelected,
  onClose,
  reloadAgents,
  editAction,
}: EditAgentsGroupsModalProps) => {
  const [selectedGroups, setSelectedGroups] = useState<Option[]>([]);
  const [finalAgents, setFinalAgents] = useState<IAgentResponse[]>([]);
  const [getAgentsStatus, setGetAgentsStatus] = useState('disabled');
  const [getAgentsError, _setGetAgentsError] = useState();
  const [saveChangesStatus, setSaveChangesStatus] = useState('disabled');
  const [isResultVisible, setIsResultVisible] = useState(false);
  const [documentResults, setDocumentResults] = useState<GroupResult[]>([]);
  const getArrayByProperty = (array: any[], propertyName: string) =>
    array.map(element => element[propertyName]);

  const handleSave = async () => {
    setIsResultVisible(true);
    setGetAgentsStatus('loading');
    setGetAgentsStatus('complete');
    setFinalAgents(selectedAgents);
    setSaveChangesStatus('loading');

    const groups = getArrayByProperty(selectedGroups, 'label');
    const agentDocumentIds = getArrayByProperty(selectedAgents, '_id');

    // Process each agent
    const processAgent = async (documentId: string) => {
      try {
        // Determine which action to perform
        const actionMethod =
          editAction === EditActionGroups.ADD
            ? getAgentManagement().addGroups
            : getAgentManagement().removeGroups;
        const result = await actionMethod(documentId, groups);

        if (!result) {
          return;
        }

        const { data, error, message } = result;
        const {
          affected_items: affectedItems,
          failed_items: failedItems,
          total_failed_items: totalFailedItems,
        } = data;

        // Update results
        setDocumentResults(results => [
          ...results,
          {
            documentId,
            result: error ? RESULT_TYPE.ERROR : RESULT_TYPE.SUCCESS,
            successAgents: affectedItems,
            errorAgents: failedItems,
            errorMessage: message,
            totalErrorAgents: totalFailedItems,
          },
        ]);
      } catch (error: any) {
        setDocumentResults(results => [
          ...results,
          {
            documentId,
            result: RESULT_TYPE.ERROR,
            errorMessage: error.message,
            errorAgents: [
              {
                error: { message: error.message },
                id: documentId,
              },
            ],
          },
        ]);
      }
    };

    try {
      await Promise.allSettled(
        agentDocumentIds.map(element => processAgent(element)),
      );
      setSaveChangesStatus('complete');
    } catch {
      setSaveChangesStatus('danger');
    } finally {
      reloadAgents();
    }
  };

  const groupsText =
    editAction === EditActionGroups.ADD ? 'Groups to add' : 'Groups to remove';

  const handleOnChangeGroupsSelect = (selectedGroups: Option[]) => {
    setSelectedGroups(selectedGroups);
  };

  const onCreateOption = (searchValue: string) => {
    const newOption = {
      label: searchValue,
    };

    // Select the option.
    setSelectedGroups([...selectedGroups, newOption]);
  };

  const selectGroupsForm = (
    <EuiForm component='form'>
      {allAgentsSelected ? (
        <EuiFormRow>
          <EuiCallOut
            color='warning'
            iconType='alert'
            title='The changes will be applied to all agents that match the filters set in the list'
          />
        </EuiFormRow>
      ) : (
        <EuiFormRow label='Selected agents'>
          <EuiText>{selectedAgents.length}</EuiText>
        </EuiFormRow>
      )}

      <EuiFormRow label={groupsText}>
        <EuiComboBox
          placeholder={groupsText}
          selectedOptions={selectedGroups}
          onChange={handleOnChangeGroupsSelect}
          isLoading={false}
          // TODO: Change when the endpoint or index pattern is available to request the groups
          // options={groups?.map(group => ({ label: group })) || []}
          noSuggestions
          onCreateOption={onCreateOption}
        />
      </EuiFormRow>
    </EuiForm>
  );

  return (
    <EuiModal onClose={onClose}>
      <EuiModalHeader>
        <EuiModalHeaderTitle>
          {editAction === EditActionGroups.ADD
            ? 'Add groups to agents'
            : 'Remove groups from agents'}
        </EuiModalHeaderTitle>
      </EuiModalHeader>
      <EuiModalBody>
        {isResultVisible ? (
          <EditAgentsGroupsModalResult
            finalAgents={finalAgents}
            editAction={editAction}
            getAgentsStatus={getAgentsStatus}
            getAgentsError={getAgentsError}
            saveChangesStatus={saveChangesStatus}
            documentResults={documentResults}
            groups={getArrayByProperty(selectedGroups, 'label')}
          />
        ) : (
          selectGroupsForm
        )}
      </EuiModalBody>
      <EuiModalFooter>
        {isResultVisible ? (
          <EuiButton
            onClick={onClose}
            fill
            disabled={
              getAgentsStatus === 'loading' || saveChangesStatus === 'loading'
            }
          >
            Close
          </EuiButton>
        ) : (
          <>
            <EuiButtonEmpty onClick={onClose}>Cancel</EuiButtonEmpty>
            <EuiButton
              onClick={handleSave}
              fill
              disabled={selectedGroups?.length === 0}
            >
              Save
            </EuiButton>
          </>
        )}
      </EuiModalFooter>
    </EuiModal>
  );
};
