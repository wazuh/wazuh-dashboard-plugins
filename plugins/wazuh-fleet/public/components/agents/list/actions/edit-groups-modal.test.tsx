import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { getAgentManagement, getToasts } from '../../../../plugin-services';
import { EditAgentGroupsModal } from './edit-groups-modal';

// Mock the plugin services
jest.mock('../../../../plugin-services', () => ({
  getAgentManagement: jest.fn(),
  getToasts: jest.fn(),
}));

describe('EditAgentGroupsModal component', () => {
  const mockAgent = {
    _source: {
      agent: {
        id: '001',
        name: 'agent1',
        groups: ['default', 'test1'],
      },
    },
  };
  const mockOnClose = jest.fn();
  const mockReloadAgents = jest.fn();
  const mockAddGroups = jest.fn().mockResolvedValue({});
  const mockRemoveGroups = jest.fn().mockResolvedValue({});
  const mockAddInfo = jest.fn();
  const mockAddDanger = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    // Setup mocks
    (getAgentManagement as jest.Mock).mockReturnValue({
      addGroups: mockAddGroups,
      removeGroups: mockRemoveGroups,
    });

    (getToasts as jest.Mock).mockReturnValue({
      addInfo: mockAddInfo,
      addDanger: mockAddDanger,
    });
  });

  test('should return the component with save disabled', async () => {
    const { container, getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={mockAgent}
        onClose={() => {}}
        reloadAgents={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();

    const agentName = getByText('agent1');

    expect(agentName).toBeInTheDocument();

    const saveButton = getByRole('button', { name: 'Save' });

    expect(saveButton).toBeInTheDocument();

    expect(saveButton).toBeDisabled();

    const cancelButton = getByRole('button', { name: 'Cancel' });

    expect(cancelButton).toBeInTheDocument();
  });

  test('should return the component', async () => {
    const { container, getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={mockAgent}
        onClose={() => {}}
        reloadAgents={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();

    const agentName = getByText('agent1');

    expect(agentName).toBeInTheDocument();

    const saveButton = getByRole('button', { name: 'Save' });

    expect(saveButton).toBeInTheDocument();

    const cancelButton = getByRole('button', { name: 'Cancel' });

    expect(cancelButton).toBeInTheDocument();
  });

  test('should add a new group', async () => {
    const { getByText, getByRole } = render(
      <EditAgentGroupsModal
        agent={mockAgent}
        onClose={mockOnClose}
        reloadAgents={mockReloadAgents}
      />,
    );
    const selectedGroup = getByRole('textbox');

    expect(selectedGroup).toBeInTheDocument();

    // Add a new group

    fireEvent.change(selectedGroup, { target: { value: 'newgroup' } });
    fireEvent.keyDown(selectedGroup, { key: 'Enter', keyCode: 13 });

    // Verify the new group appears in the UI
    expect(getByText('newgroup')).toBeInTheDocument();

    const saveButton = getByRole('button', { name: 'Save' });

    expect(saveButton).toBeInTheDocument();
    expect(saveButton).not.toBeDisabled();

    // Click save button
    await act(async () => {
      fireEvent.click(saveButton);
    });

    // Verify the mocks were called
    expect(mockAddGroups).toHaveBeenCalledWith('001', ['newgroup']);
    expect(mockAddInfo).toHaveBeenCalledWith({
      title: 'Agent groups edited',
      text: 'Agent agent1 groups have been updated',
    });
    expect(mockReloadAgents).toHaveBeenCalled();
    expect(mockOnClose).toHaveBeenCalled();
  });

  test('should remove a group', async () => {
    const { getByTitle, getByRole } = render(
      <EditAgentGroupsModal
        agent={mockAgent}
        onClose={mockOnClose}
        reloadAgents={mockReloadAgents}
      />,
    );
    const removeButton = getByTitle(
      'Remove default from selection in this group',
    );
    const saveButton = getByRole('button', { name: 'Save' });

    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toBeDisabled();
    expect(removeButton).toBeInTheDocument();
    // Click remove button
    await act(async () => {
      fireEvent.click(removeButton);
    });

    expect(saveButton).not.toBeDisabled();

    // Click save button
    await act(async () => {
      fireEvent.click(saveButton);
    });
    // Verify the mocks were called
    expect(mockRemoveGroups).toHaveBeenCalledWith({
      agentId: '001',
      groupIds: ['default'],
    });
    expect(mockAddInfo).toHaveBeenCalledWith({
      title: 'Agent groups edited',
      text: 'Agent agent1 groups have been updated',
    });
    expect(mockReloadAgents).toHaveBeenCalled();
    expect(mockOnClose).toHaveBeenCalled();
  });

  test('should close modal when cancel button is clicked', async () => {
    const { getByRole } = render(
      <EditAgentGroupsModal
        agent={mockAgent}
        onClose={mockOnClose}
        reloadAgents={mockReloadAgents}
      />,
    );
    const cancelButton = getByRole('button', { name: 'Cancel' });

    expect(cancelButton).toBeInTheDocument();
    // Click cancel button
    await act(async () => {
      fireEvent.click(cancelButton);
    });
    // Verify onClose was called
    expect(mockOnClose).toHaveBeenCalled();
    expect(mockReloadAgents).not.toHaveBeenCalled();
    expect(mockRemoveGroups).not.toHaveBeenCalled();
    expect(mockAddGroups).not.toHaveBeenCalled();
  });
});
