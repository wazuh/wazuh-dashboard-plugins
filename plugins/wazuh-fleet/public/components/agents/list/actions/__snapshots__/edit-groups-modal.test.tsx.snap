// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditAgentGroupsModal component should return the component 1`] = `<div />`;

exports[`EditAgentGroupsModal component should return the component with save disabled 1`] = `<div />`;
